### PR TITLE
[RFC] Add cmdline mode to modechange of RPC and tests

### PR DESCRIPTION
**Pull request by chemzqm**
I can't make sure the test is correct, when I run the test command

```
TEST_FILE=test/functional/ui/highlight_spec.lua   make unittest
```
The command get suspended.
screen shot 2016-11-26 at 8 03 06 am

### CODE REVIEW

**Comment by justinmk** [comment]
> TEST_FILE=test/functional/ui/highlight_spec.lua   make unittest

Test command should be:

    TEST_FILE=test/functional/ui/screen_basic_spec.lua make functionaltest

(`unittest` is a different set of tests than `functionaltest`)

**Comment by chemzqm** [comment]
I tried `TEST_FILE=test/functional/ui/screen_basic_spec.lua make functionaltest` still not works, looks like something is missing.

**Comment by brcolow** [comment]
You can utilize the [Travis CI](https://travis-ci.org/neovim/neovim/builds/178968337), [Appveyor](https://ci.appveyor.com/project/neovim/neovim-1kmqj/build/9455), and [QuickBuild](http://neovim-qb.szakmeister.net/build/9405) results to see the errors.

**Comment by chemzqm** [comment]
I've checked the log https://travis-ci.org/neovim/neovim/jobs/178968346, looks like the test not runned.

**Comment by bfredl** [comment]
Also need this change:
```diff
diff --git a/test/functional/ui/screen.lua b/test/functional/ui/screen.lua
index 6c1a178..f130841 100644
--- a/test/functional/ui/screen.lua
+++ b/test/functional/ui/screen.lua
@@ -374,7 +374,8 @@ function Screen:_handle_mouse_off()
 end
 
 function Screen:_handle_mode_change(mode)
-  assert(mode == 'insert' or mode == 'replace' or mode == 'normal')
+  assert(mode == 'insert' or mode == 'replace'
+         or mode == 'normal' or mode == 'cmdline')
   self.mode = mode
 end
```

then you will some futher errors for the actual tests, like the cursor position is wrong:

```diff
diff --git a/test/functional/ui/screen_basic_spec.lua b/test/functional/ui/screen_basic_spec.lua
index 6989daf..6032489 100644
--- a/test/functional/ui/screen_basic_spec.lua
+++ b/test/functional/ui/screen_basic_spec.lua
@@ -646,11 +646,11 @@ describe('Screen', function()
     it('works in cmdline mode', function()
       feed(':')
       screen:expect([[
-        ^                         |
+                                 |
         {0:~                        }|
         {0:~                        }|
         {0:~                        }|
-        :                        |
+        :^                        |
       ]],nil,nil,function ()
         eq("cmdline", screen.mode)
       end)
```

**Comment by bfredl** [comment]
Also `tui_mode_change` in `src/nvim/tui/tui.c` needs to be updated. I think the best would be use `set_cursor_shape_bar` just like insert mode.

**Comment by chemzqm** [comment]
Tried to make the test get passed on my mac.

File: src/nvim/ui.c [hunk 1]
```diff
@@ -537,6 +537,8 @@ static void ui_mode_change(void)
     mode = REPLACE;
   else if (State & INSERT)
     mode = INSERT;
+  else if (State & CMDLINE)
```
Comment (bfredl) [inline src/nvim/ui.c]: you need to change this to use braces (the entire if/else if/...) for the linter to pass.

Comment (chemzqm) [inline src/nvim/ui.c]: @bfredl fixed

File: src/nvim/ui.c [hunk 2]
```diff
@@ -533,12 +533,15 @@ static void ui_mode_change(void)
     return;
   }
   /* Get a simple UI mode out of State. */
```
Comment (bfredl) [inline src/nvim/ui.c]: This needs to changed to new-style comment `//`

**Comment by bfredl** [comment]
One more linter fix needed, then I think this is ready.

**Comment by Chad-** [comment]
Could this be expanded to add more states? In particular `VISUAL` would be nice to have, and `OP_PENDING` would be useful for avoiding #5685 until it is fixed.

**Comment by bfredl** [comment]
Sure, but let's do that in a follow-up PR.

File: src/nvim/ui.c (hunk 2 above)
Comment (chemzqm) [inline src/nvim/ui.c]: Is squash git commits needed?

**Comment by bfredl** [comment]
> Is squash git commits needed?

Please squash them to one commit

**Comment by chemzqm** [comment]
@bfredl It's ready

**Comment by bfredl** [comment]
failures look unrelated, merging.

**Pull request merged**